From 9c8cc0306a676a0f6086a03c8c33b4525cfa3e85 Mon Sep 17 00:00:00 2001
From: Gustavo Duarte <gduarte@redhat.com>
Date: Mon, 5 Jan 2015 21:30:09 -0500
Subject: [powerpc] Fill in si_addr_lsb siginfo field

Message-id: <1420493409-15884-3-git-send-email-gduarte@redhat.com>
Patchwork-id: 102466
O-Subject: [RHEL7.1 PATCH BZ 1173267 2/2] powerpc: Fill in si_addr_lsb siginfo field
Bugzilla: 1173267
RH-Acked-by: Rik van Riel <riel@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>

BZ 1173267
Upstream Status: 9d57472f61acd7c3a33ebf5a79361e316d8ffbef

commit 9d57472f61acd7c3a33ebf5a79361e316d8ffbef
Author: Anton Blanchard <anton@samba.org>
Date:   Wed Sep 24 16:59:58 2014 +1000

    powerpc: Fill in si_addr_lsb siginfo field

    Fill in the si_addr_lsb siginfo field so the hwpoison code can
    pass to userspace the length of memory that has been corrupted.

    Signed-off-by: Anton Blanchard <anton@samba.org>
    Signed-off-by: Michael Ellerman <mpe@ellerman.id.au>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/mm/fault.c b/arch/powerpc/mm/fault.c
index 1caee83..e4717af 100644
--- a/arch/powerpc/mm/fault.c
+++ b/arch/powerpc/mm/fault.c
@@ -33,6 +33,7 @@
 #include <linux/magic.h>
 #include <linux/ratelimit.h>
 #include <linux/context_tracking.h>
+#include <linux/hugetlb.h>
 
 #include <asm/firmware.h>
 #include <asm/page.h>
@@ -118,6 +119,7 @@ static int do_sigbus(struct pt_regs *regs, unsigned long address,
        unsigned int fault)
 {
  siginfo_t info;
+ unsigned int lsb = 0;
 
  up_read(&current->mm->mmap_sem);
 
@@ -133,7 +135,13 @@ static int do_sigbus(struct pt_regs *regs, unsigned long address,
     current->comm, current->pid, address);
    info.si_code = BUS_MCEERR_AR;
   }
+
+  if (fault & VM_FAULT_HWPOISON_LARGE)
+   lsb = hstate_index_to_shift(VM_FAULT_GET_HINDEX(fault));
+  if (fault & VM_FAULT_HWPOISON)
+   lsb = PAGE_SHIFT;
 #endif
+  info.si_addr_lsb = lsb;
   force_sig_info(SIGBUS, &info, current);
   return MM_FAULT_RETURN;
  }
-- 
1.7.1